redash/models.py

```diff
@@ -348,12 +349,32 @@ def __unicode__(self):
         return unicode(self.id)
 
 
+class EmailType(TypeDecorator):
+    """
+    A case-insensitive string
+    """
+    impl = db.String
+    comparator_factory = CaseInsensitiveComparator
```

Comment:
Assuming that all emails already lower cased, why do we need a special comparator?

Reply:
The custom comparator is to handle when inputs from anywhere (api, cli) won't need to specify `.lower()` when performing equality comparison.

Here's an example of how the SQL would look, with and without the custom comparator:

```python
>>> from redash.models import User
>>> user1 = User.query.filter(User.email == "TEST@example.com")
>>> print(user1)
```

```sql
-- Before
SELECT users.groups AS users_groups, users.updated_at AS users_updated_at, users.created_at AS users_created_at, users.id AS users_id, users.org_id AS users_org_id, users.name AS users_name, users.email AS users_email, users.password_hash AS users_password_hash, users.api_key AS users_api_key
FROM users
WHERE users.email = %(email_1)s

-- After, custom comparator uses SQL `lower`
SELECT users.groups AS users_groups, users.updated_at AS users_updated_at, users.created_at AS users_created_at, users.id AS users_id, users.org_id AS users_org_id, users.name AS users_name, users.email AS users_email, users.password_hash AS users_password_hash, users.api_key AS users_api_key
FROM users
WHERE lower(users.email) = lower(%(lower_1)s)
```

So this prevents us needing to handle every case of anyone trying to compare against this column by ensuring the underlying SQL handles it.